File: Students/anastasiyaChabotska/day_3/lc142.js
// Problem Statement #

// Given the head of a Singly LinkedList that contains a cycle, write a function to find the starting node of the cycle.

class Node {
  constructor(value, next=null){
    this.value = value;
    this.next = next;
  }
}

const find_cycle_start = function(head){

  let slow = head;
  let fast = head;

  while (fast && fast.next){
    fast = fast.next.next;
    slow = slow.next;

    if(slow == fast)
    break;
  }

  //traced the logic on the paper, looks like the start of the cycle will always be the slow.next
  //tested with other test cases, slow.next is incorrect
  //some smarter math is involved
  //we can count the length of the cycle, but what can it give us? 
  //going backwards, until pointers meet? but it's singly linked...

  return slow.next;
};


head = new Node(1)
head.next = new Node(2)
head.next.next = new Node(3)
head.next.next.next = new Node(4)
head.next.next.next.next = new Node(5)
head.next.next.next.next.next = new Node(6)

head.next.next.next.next.next.next = head.next.next
console.log(`LinkedList cycle start: ${find_cycle_start(head).value}`)

head.next.next.next.next.next.next = head.next.next.next
console.log(`LinkedList cycle start: ${find_cycle_start(head).value}`)

head.next.next.next.next.next.next = head

File: Students/Xin_Jin/lc142.java
package BootCamp;

import java.util.HashSet;
import java.util.Set;

class ListNode {
      int val;
      ListNode next;
      ListNode(int x) {
          val = x;
          next = null;
      }
  }
public class lc142 {
    // Given the head of a Singly LinkedList that contains a cycle, write a function to find the starting node of the cycle.
    public static void main(String[] args) {
        ListNode node=new ListNode(3);
        node.next=new ListNode(2);
        node.next.next=new ListNode(0);
        node.next.next.next=new ListNode(-4);
        node.next.next.next.next=node.next;
        System.out.println(findStartingNode(node).val);
    }
    public static ListNode findStartingNode(ListNode head)
    {
        Set<ListNode> set=new HashSet<>();
        int index=0;
        while(head!=null)
        {
            if(set.contains(head))return head;
            set.add(head);
            index++;
            head=head.next;
        }
        return null;
    }
}
